rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  flake. (For example, `fetchGit` no longer fails if we already have
+  the specified repository and `ref`.)
+
+* Fetched flakes are registered as garbage collector roots, so running
+  the garbage collector while on an airplane will not ruin your day.
+
+* There is an option `--no-net` to explicitly prevent updating the
+  registry or doing other things that need the network (such as
+  substituting). This is the default if there are no configured
+  non-loopback network interfaces.
+
+## CI integration
+
+Flakes are useful for CI systems since `flake.nix` and the lock file
+exactly specify the input repositories. For Hydra this means that you
+no longer need to configure any jobset inputs; just a top-level flake
+reference is enough. For example, the [configuration of the `patchelf`
+jobset](https://hydra.nixos.org/jobset/patchelf/master#tabs-configuration)
+just points to `github:NixOS/patchelf` and doesn't need to specify
+`nixpkgs`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure evaluation breaks certain workflows. In particular, it breaks the
+use of the Nixpkgs configuration file. Similarly, there are people who
+rely on `$NIX_PATH` to pass configuration data to NixOps
+configurations.
+
+# Alternatives
+[alternatives]: #alternatives
+
+For composition of multi-repository projects, the main alternative is
+to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
+in other repositories. However, since there is no explicit listing of
+dependencies, this does not provide automatic updating (i.e. there is
+no equivalent of `nix flake update`).
+
+Instead of a `flake.nix`, flakes could store their metadata in a
+simpler format such as JSON or TOML. This avoids the Turing tarpit
+where getting flake metadata requires the execution of an arbitrarily
+complex, possibly non-terminating program.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* How to handle the system type? Currently `x86_64-linux` is
+  hard-coded everywhere.
+
+* How to do Nixpkgs overlays? In principle, overlays would just be a
+  kind of flake output.
+
+* More in general, how to handle flake arguments? This must be done in
+  a way that maintains hermetic evaluation and evaluation caching.
+
+* What are the criteria for inclusion in the global flake registry?
+
+* Hammer out the details of NixOS/NixOps support for flakes.
+
+* Currently, if flake dependencies (repositories or branches) get
+  deleted, rebuilding the flake may fail. (This is similar to
+  `fetchurl` referencing a stale URL.) We need a command to gather all
+  flake dependencies and copy them somewhere else (possibly vendor
+  them into the repository of the calling flake).
+
+# Future work
+[future]: #future-work
+
+* The "edition" feature enables future Nix changes, including language
+  changes. For example, changing the parsing of multiline strings
+  (https://github.com/NixOS/nix/pull/2490) could be conditional on the
+  flake's edition.
```

Comment:
This is a good practice. Rust does something similar with [editions](https://doc.rust-lang.org/nightly/edition-guide/editions/index.html).

The key here is that the compiler (or evaluator in nix’s case) never loses the ability to work with older editions.

Reply:
How would parsing these work? With Cargo, it’s not like the TOML format is changing between editions. Does the evaluator have to try evaluating with each version of the syntax, starting with the newest, until it finds one that works?

Reply:
Good question, but do we really want to change the nix syntax in a backwards-incompatible way at this point? I think probably no.

If we don’t intend to change the syntax it’s just about the semantics of the fields, which is what `editions` are about I presume.

Reply:
> Good question, but do we really want to change the nix syntax in a backwards-incompatible way at this point? I think probably no.

There's https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d

Reply:
@alyssais That's a good point, and is an argument in favor of using JSON/TOML for `flake.nix`.

Reply:
Shouldn't `edition` just be evaluated lazily? That seems like a sane requirement.

Reply:
It can't be evaluated lazily. It has to be evaluated eagerly because the interpretation of all other attributes (potentially) depends on it.

Reply:
I guess technically there is also the option of including edition into the main flake file name, like some codebases used `.py2` and `.py3` extensions at some point.